src/main/java/net/minecraftforge/event/world/JigsawPatternInitEvent.java

```diff
@@ -0,0 +1,285 @@
+     * @param pieces A list of JigsawPieces and weights that should be added to the pool
+     */
+    public void addBuildings(@Nonnull ResourceLocation category, @Nonnull List<Pair<JigsawPiece, Integer>> pieces)
+    {
+        this.newBuildings.computeIfAbsent(category, categoryName -> Lists.newArrayList()).addAll(pieces);
+    }
+
+    /**
+     * removes single JigsawPiece from a JigsawPattern
+     * @param category The category the building should be added. If null the default one will be selected. Some {@link JigsawPattern} may not contains the default category.
+     * @param building The identifier of the JigsawPiece {@link JigsawPiece#getRegistryName()} (see e.g. {@link net.minecraft.world.gen.feature.structure.PlainsVillagePools}
+     */
+    public void removeBuilding(@Nonnull ResourceLocation category, @Nonnull ResourceLocation building)
+    {
+        this.remove.computeIfAbsent(category, categoryName -> Lists.newArrayList()).add(building);
+    }
+
+    /**
+     * removes multiple JigsawPieces from a JigsawPattern
+     * @param category The category the building should be added. If null the default one will be selected. Some {@link JigsawPattern} may not contains the default category.
+     * @param buildings The identifiers of the JigsawPieces {@link JigsawPiece#getRegistryName()} (see e.g. {@link net.minecraft.world.gen.feature.structure.PlainsVillagePools}
+     */
+    public void removeBuildings(@Nonnull ResourceLocation category, @Nonnull List<ResourceLocation> buildings)
+    {
+        remove.computeIfAbsent(category, categoryName -> Lists.newArrayList()).addAll(buildings);
+    }
+
+    public List<JigsawCategory> getPool()
+    {
+        for(ResourceLocation remove: this.removeCategories)
+        {
+            if(this.jigsawPieces.remove(remove) == null){
+                LOGGER.debug("Tried to remove nonexistent JigsawCategory {} from JigsawPattern {}", remove, this.jigsawPoolRegistryName);
+            }
+        }
+        this.jigsawPieces.putAll(this.newCategories);
+        for (Map.Entry<ResourceLocation,List<ResourceLocation>> remove: this.remove.entrySet())
+        {
+            if(this.jigsawPieces.containsKey(remove.getKey())){
+                this.jigsawPieces.get(remove.getKey()).getPieces().removeIf(jigsawPieceIntegerPair -> remove.getValue().contains(jigsawPieceIntegerPair.getFirst().getRegistryName()));
+            }
+        }
+        for(Map.Entry<ResourceLocation,List<Pair<JigsawPiece,Integer>>> entry: this.newBuildings.entrySet())
+        {
+            JigsawCategory category = this.jigsawPieces.get(entry.getKey());
+            if(category == null){
+                LOGGER.debug("Building cannot be added. JigsawCategory {} doesn't exits or were removed in JigsawPattern {}", entry.getKey(), this.jigsawPoolRegistryName);
```

Comment:
I would check if it was actually removed (so in the remove set). If yes, log at debug, else log at warn